### DISCUSSION TITLE
how to get total issues count of repository from github api ?

### DISCUSSION THREAD

**Discussion by 0xabdulkhaliq** [community/community#61508] [Apps, API and Webhooks] [answered]
### Select Topic Area

Question

### Hey there Everyone 👋🏻

Recently i tried to fetch a repository's total issues count without getting those issues itself, i tried many methods to get that count but in the end i got frustated.

This is what i get after calling issues
```
0: Object { url: "https://api.github.com/repos/pranjay-poddar/Dev-Geeks/issues/3986", repository_url: "https://api.github.com/repos/pranjay-poddar/Dev-Geeks", labels_url: "https://api.github.com/repos/pranjay-poddar/Dev-Geeks/issues/3986/labels{/name}", … }
```
But when we look at the repository's issue block we can see the numbers are indicated like this

![Screenshot 2023-07-22 at 13-58-38 Issues · pranjay-poddar_Dev-Geeks](https://github.com/community/community/assets/92252895/64086c21-852c-47b8-9503-98819ac299a8)
![Screenshot 2023-07-22 at 13-59-11 Issues · pranjay-poddar_Dev-Geeks](https://github.com/community/community/assets/92252895/247835d9-3b41-482b-9482-a2e6e0388118)

I don't get this about **how this site only getting the count perfectly but i can't get these values from the API ?**

I tried to used several methods on javascript like `filter`, `map` to count them but it takes so much of time and resource to do that. It ends with duplicate values also.

I want to know that the API has the support to just return the total count of issues and PR's.

Your thoughts and feedbacks are highly appreciated!

Thank you for your attention regarding this matter.

**Comment by LangLangBart** [comment]
- try it with the `search` from the `REST API`
  - [GitHub Docs - Search](https://docs.github.com/en/rest/search/search?apiVersion=2022-11-28#search-issues-and-pull-requests)

```sh
gh api /search/issues?q="repo:pranjay-poddar/Dev-Geeks+is:issue" --jq .total_count
# 2131
gh api /search/issues?q="repo:pranjay-poddar/Dev-Geeks+is:issue+is:closed" --jq .total_count
# 1521
gh api /search/issues?q="repo:pranjay-poddar/Dev-Geeks+is:issue+is:open" --jq .total_count
# 610
```

**Reply by LangLangBart** [reply]
- alternatively, the `GraphQL API` can output the total count too
  - [GitHUb Docs - GraphQL](https://docs.github.com/en/graphql/reference/objects#repository)
```graphql
query ($owner: String = "pranjay-poddar", $name: String = "Dev-Geeks") {
  repository(owner: $owner, name: $name) {
    all:issues {
      totalCount
    }
    closed:issues(states:CLOSED) {
      totalCount
    }
    open:issues(states:OPEN) {
      totalCount
    }
  }
}
```

```json
{
  "data": {
    "repository": {
      "all": {
        "totalCount": 2131
      },
      "closed": {
        "totalCount": 1521
      },
      "open": {
        "totalCount": 610
      }
    }
  }
}
```

**Reply by 0xabdulkhaliq** [reply]
Wow @LangLangBart, i have tried using Graph QL and it worked thank you for that!

But the thing i can't understand that i can't get the contributors count, i tried collabrators but it returns null

![image](https://github.com/community/community/assets/92252895/537534da-2203-44cc-a534-92e5de916c1b)

Can you resolve this ?

**Reply by LangLangBart** [reply, marked as answer]
>  i can't get the contributors count, i tried collabrators but it returns null

- **NOTE:** GitHub’s `REST API` and `GraphQL API` have no parity, they are a Venn Diagram, feature-wise.
  - the `GraphQL API`  has no `contributors` object, only a `collabrators` object and this info is only accessible when you have some kind of access to the repo, you should receive an error message 
  - `collaborators` != `contributors`
    - [GitHub Docs - list-repository-collaborators](https://docs.github.com/en/rest/collaborators/collaborators?apiVersion=2022-11-28#list-repository-collaborators)
    - [GitHub Docs -  list-repository-contributors](https://docs.github.com/en/rest/repos/repos?apiVersion=2022-11-28#list-repository-contributors) 

```graphql
{
  repository(owner: "pranjay-poddar", name: "Dev-Geeks") {
    collaborators {
      totalCount
    }
  }
}
```

```json
{
  "data": {
    "repository": {
      "collaborators": null
    }
  },
  "errors": [
    {
      "type": "FORBIDDEN",
      "path": [
        "repository",
        "collaborators"
      ],
      "locations": [
        {
          "line": 3,
          "column": 5
        }
      ],
      "message": "You do not have permission to view repository collaborators."
    }
  ]
}
```

- the only object that might be of interest is the `mentionableUsers` in the `GraphQL`. I don't know why it's called that, but the names you get from there are very similar to the list of names from the `Contributors` in the `REST API`.

---

- checking the difference between the names returned from the `REST API` for `contributors` vs the  `GraphQL API` for `mentionableUsers` below

```sh
# REST API
gh api --paginate repos/pranjay-poddar/Dev-Geeks/contributors --jq '.[].login' | sort >rest_api.txt

gh api --paginate repos/pranjay-poddar/Dev-Geeks/contributors --jq '.[].login' | wc -l
# 405

# GraphQL API
gh api graphql --paginate \
	--raw-field query=$'query ($endCursor: String){
		repository(owner: "pranjay-poddar", name: "Dev-Geeks"){
			mentionableUsers(first: 100 after: $endCursor){
				nodes { login }
				pageInfo { hasNextPage endCursor }
				totalCount }}}' \
	--jq '.data.repository.mentionableUsers.nodes[].login' | sort >graphql_api.txt

gh api graphql \
	--raw-field query=$'{repository(owner: "pranjay-poddar", name: "Dev-Geeks"){
			mentionableUsers{totalCount}}}' \
	--jq '.data.repository.mentionableUsers.totalCount'
# 407

# the differences between the lists
comm -3 graphql_api.txt rest_api.txt
# MohitGupta121
# UditSharma9999
#        codesee-maps[bot]
#        dependabot[bot]
# varunlmxd
# yvrjprshr
```

**Reply by LangLangBart** [reply]
The Web UI shows 409 `Contributors`, whereas the `GraphQL` API only returns 407 without counting `codesee-maps[bot]`  and `dependabot[bot]`.

----

**NOTE**: Another user reported a big difference between the total number `mentionableUsers` for the `GraphQL API` and the `contributors` number from the `REST API` - #53747 (26/Apr/23). 

The `mentionableUsers` object might work for this specific repo in your example, but the description indicates a different purpose.  

> _mentionableUsers: A list of Users that can be mentioned in the context of the repository.
> Source: [GitHub Docs - GraphQL API - objects#repository ](https://docs.github.com/en/graphql/reference/objects#repository)_

**Reply by 0xabdulkhaliq** [reply]
Glad to get this detailed breakdown about `mentionableUsers` 🖖🏻 

Thank you so much once again!

**Reply by LangLangBart** [reply]
> Thank you so much once again!

👍 

---

PS: in your posted image, the number of `Closed Pull Requests:` is at `1267`, but that includes only merged pull requests.


```graphql
query ($owner: String = "pranjay-poddar", $name: String = "Dev-Geeks") {
  repository(owner: $owner, name: $name) {
    all: pullRequests {
      totalCount
    }
    open: pullRequests(states: OPEN) {
      totalCount
    }
    closed_merged: pullRequests(states: [CLOSED, MERGED]) {
      totalCount
    }
    merged: pullRequests(states: MERGED) {
      totalCount
    }
    closed: pullRequests(states: CLOSED) {
      totalCount
    }
  }
  __type(name: "PullRequestState") {
    description
    enumValues {
      name
      description
    }
  }
}
```


```json
{
  "data": {
    "repository": {
      "all": {
        "totalCount": 1884
      },
      "open": {
        "totalCount": 115
      },
      "closed_merged": {
        "totalCount": 1769
      },
      "merged": {
        "totalCount": 1274
      },
      "closed": {
        "totalCount": 495
      }
    },
    "__type": {
      "description": "The possible states of a pull request.",
      "enumValues": [
        {
          "name": "OPEN",
          "description": "A pull request that is still open."
        },
        {
          "name": "CLOSED",
          "description": "A pull request that has been closed without being merged."
        },
        {
          "name": "MERGED",
          "description": "A pull request that has been closed by being merged."
        }
      ]
    }
  }
}
```

**Reply by 0xabdulkhaliq** [reply]
Yep, i want only the merged ones so that we can properly count the PR's which are contributed.

By the way thanks for sharing this code with such as amount of detail, I hope this will be useful in any other scenarios!

**Comment by Arnab0x** [comment]
@LangLangBart Hi thanks for the detailed explanation! will this work if the number of issues (open or closed) is huge? or are there any limits like the rest api (number of items per page or pagination)?
